module/VuFind/src/VuFind/I18n/Translator/TranslatorAwareTrait.php

```diff
@@ -260,4 +265,21 @@ protected function extractTextDomain($target)
         }
         return ['default', is_array($target) ? $parts[0] : $target];
     }
+
+    /**
+     * Make sure there are not any illegal characters in the translation key
+     * that might prevent successful lookup in language files.
+     *
+     * @param string $key Key to sanitize
+     *
+     * @return string Sanitized key
+     */
+    protected function sanitizeTranslationKey(string $key): string
+    {
+        // The characters ()!? are not allowed in keys in the Lokalise translation
+        // platform, so they should not be allowed in our code. We'll replace them
+        // with underscores so that translations can still be provided if the input
+        // cannot be changed (e.g. if it comes from a third-party system).
+        return preg_replace('/[()!?]/', '_', $key);
```

Comment:
Could we just use str_replace to avoid much of the performance penalty with regex?

Reply:
Done! I was nervous about multi-byte support, but it looks like that's not really an issue in this case.